### PR TITLE
重命名依赖工程中R类名为主lib的R类名

### PR DESCRIPTION
**Pull request by aprz512**
添加可选项，使用重命名R文件的方式替换原来生成R文件的方式

### CODE REVIEW

**Comment by kezong** [comment]
Nice! It's great job. 
Thanks for @aasitnikov @aprz512 #184 #240

**Comment by kezong** [comment]
I'll merge first and then make some optimization.

File: source/src/main/java/com/kezong/fataar/RTransform.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+ * |-- com.lib1.R
+ * |-- com.lib2.R
+ * <p>
+ * rename com.lib1.R and com.lib2.R to com.sdk.R
+ */
+public class RTransform extends Transform {
+
+    private static final int CPU_COUNT = Runtime.getRuntime().availableProcessors();
+
+    private final ExecutorService executor = Executors.newFixedThreadPool(CPU_COUNT + 1);
```
Comment (aasitnikov) [inline source/src/main/java/com/kezong/fataar/RTransform.java:52]: In Gradle there's [Worker API](https://docs.gradle.org/nightly/userguide/worker_api.html) for splitting task action into smaller actions, better to use it instead

File: source/src/main/java/com/kezong/fataar/RTransform.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                future.get();
+            } catch (ExecutionException e) {
+                e.printStackTrace();
+            }
+        }
+
+        futures.clear();
+
+        long endTime = System.currentTimeMillis();
+        System.out.println("[fat-aar]Task :transformClassesWithRenameRFor"
```
Comment (aasitnikov) [inline source/src/main/java/com/kezong/fataar/RTransform.java:145]: I think using `getLogger()` is better than just `System.out.println`

File: source/src/main/java/com/kezong/fataar/RTransform.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                + (endTime - startTime)
+                + "ms");
+    }
+
+    private Map<String, String> buildTransformTable(String variantName) {
+        final List<String> resourceTypes = Arrays.asList("anim", "animator", "array", "attr", "bool", "color", "dimen",
+                "drawable", "font", "fraction", "id", "integer", "interpolator", "layout", "menu", "mipmap", "plurals",
+                "raw", "string", "style", "styleable", "transition", "xml");
+
+        Collection<AndroidArchiveLibrary> libraries = VariantDependenciesStore.getLibraries(variantName);
```
Comment (aasitnikov) [inline source/src/main/java/com/kezong/fataar/RTransform.java:157]: I think using static to pass properties from plugin to task is a bad practice - better to use Gradle's Property API

File: source/src/main/groovy/com/kezong/fataar/FatLibraryPlugin.groovy [hunk 4]
```diff
@@ -204,4 +206,24 @@ class FatLibraryPlugin implements Plugin<Project> {
             }
         }
     }
+
+    private void registerTransform() {
+        File manifest = project.android.sourceSets.main.manifest.srcFile
+        String packageName = new XmlParser().parse(manifest).@package
+        // 在 project.afterEvaluate 内部注册不生效
+        // 不在 project.afterEvaluate 内部获取不到扩展属性值
```
Comment (aasitnikov) [inline source/src/main/groovy/com/kezong/fataar/FatLibraryPlugin.groovy:214]: Could you please use English in vastly used open-source library?

File: source/src/main/groovy/com/kezong/fataar/FatLibraryPlugin.groovy [hunk 5]
```diff
@@ -204,4 +206,24 @@ class FatLibraryPlugin implements Plugin<Project> {
...
+        project.android.registerTransform(new RTransform(project, packageName))
+
+        project.afterEvaluate {
+            project.android.libraryVariants.all { variant ->
+                String variantName = variant.name.capitalize()
+                def transformTask = project.tasks.getByName("transformClassesWithRenameRFor${variantName}")
+                if (transformTask) {
+                    transformTask.onlyIf {
+                        project.fatLibs.renameR
+                    }
```
Comment (aasitnikov) [inline source/src/main/groovy/com/kezong/fataar/FatLibraryPlugin.groovy:224]: I'm not sure if AGP will use proper classes if transform task will be disabled

File: source/src/main/groovy/com/kezong/fataar/FatLibraryPlugin.groovy [hunk 6]
```diff
@@ -204,4 +206,24 @@ class FatLibraryPlugin implements Plugin<Project> {
...
+        File manifest = project.android.sourceSets.main.manifest.srcFile
+        String packageName = new XmlParser().parse(manifest).@package
+        // 在 project.afterEvaluate 内部注册不生效
+        // 不在 project.afterEvaluate 内部获取不到扩展属性值
+        project.android.registerTransform(new RTransform(project, packageName))
+
+        project.afterEvaluate {
+            project.android.libraryVariants.all { variant ->
+                String variantName = variant.name.capitalize()
+                def transformTask = project.tasks.getByName("transformClassesWithRenameRFor${variantName}")
```
Comment (aasitnikov) [inline source/src/main/groovy/com/kezong/fataar/FatLibraryPlugin.groovy:220]: As I wrote in [comment](https://github.com/kezong/fat-aar-android/issues/184#issuecomment-682585942), this task should depend on `mergeClasses` task

File: source/src/main/java/com/kezong/fataar/RTransform.java (hunk 3 above)
Comment (kezong) [reply source/src/main/java/com/kezong/fataar/RTransform.java:157]: I haven't found a good way to get the package name of the submodule. Do you have any suggestions?

Comment (aasitnikov) [reply source/src/main/java/com/kezong/fataar/RTransform.java:157]: I don't see any other way, but to parse AndroidManifest.xml of submodule. Plugin can pass `Set<File>` through property, where each file corresponds to submodule manifest, and then `RTransform` can parse manifests for actual package names inside `transform()` - that way no additional I/O would be done during configuration time.

**Pull request merged**